### PR TITLE
[lightning_base] fix s2s logging, only make train_loader once

### PR DESCRIPTION
**Pull request by sshleifer**
setup is called many times (incl twice by trainer.test), creating a dataloader each time. Will only creating a train_loader on the first call cause bad side effects that I don't understand @nateraw @williamFalcon ?

I read docs [https://pytorch-lightning.readthedocs.io/en/latest/introduction_guide.html], so I think I'm fine, but not sure.

cc @stas00

Also:
- add a fast test for run_ner
- [pl] centralize `data_dir` argument to `add_generic_args` cause rule of 3

Checks:
- verified xsum distillation trains well, has good LR logs. (warmup+linear decay are honored.)

### CODE REVIEW

File: examples/token-classification/test_ner_examples.py [hunk 1]
```diff
@@ -31,3 +34,24 @@ def test_run_ner(self):
         with patch.object(sys, "argv", ["run.py"] + testargs):
             result = run_ner.main()
             self.assertLess(result["eval_loss"], 1.5)
+
+
+    def test_run_ner_pl(self):
```
Comment (sshleifer) [inline examples/token-classification/test_ner_examples.py:39]: this doesn't test lightning and should be in a separate PR.

File: examples/lightning_base.py [hunk 2]
```diff
@@ -149,15 +151,26 @@ def test_step(self, batch, batch_nb):
...
-        self.train_loader = dataloader
-        self.total_steps = (
-            (len(dataloader.dataset) // (train_batch_size * max(1, self.hparams.gpus)))
-            // self.hparams.accumulate_grad_batches
-            * float(self.hparams.max_epochs)
-        )
+    @property
+    def total_steps(self) -> int:
+        """The number of total training steps that will be run. Used for lr scheduler purposes."""
+        if not hasattr(self, 'train_loader'):
```
Comment (sshleifer) [inline examples/lightning_base.py:157]: this way, we only make it once.
I want to pass shuffle=True, but I'm scared.

File: examples/text-classification/run_pl_glue.py [hunk 3]
```diff
@@ -161,13 +161,6 @@ def add_model_specific_args(parser, root_dir):
             type=int,
             help="The number of GPUs allocated for this, it is by default 0 meaning none",
         )
-        parser.add_argument(
```
Comment (sshleifer) [inline examples/text-classification/run_pl_glue.py:164]: centralize data_dir argument to `add_generic_args` cause rule of 3

File: examples/lightning_base.py [hunk 4]
```diff
@@ -149,15 +149,25 @@ def test_step(self, batch, batch_nb):
...
-    def setup(self, step):
-        train_batch_size = self.hparams.train_batch_size
-        dataloader = self.get_dataloader("train", train_batch_size)
-        self.train_loader = dataloader
-        self.total_steps = (
-            (len(dataloader.dataset) // (train_batch_size * max(1, self.hparams.gpus)))
-            // self.hparams.accumulate_grad_batches
-            * float(self.hparams.max_epochs)
-        )
+    @staticmethod
```
Comment (sshleifer) [inline examples/lightning_base.py:152]:
```suggestion
```
